private/bufpkg/bufimage/bufimagebuild/builder.go

```diff
@@ -125,9 +130,118 @@ func (b *builder) build(
+		directDepsIdentities[directDep.IdentityString()] = "explicit"
+	}
+	// if it's a workspace build, and there's any named module in it, then it's considered a direct
+	// dependency too
+	if localWorkspace != nil {
+		for _, mod := range localWorkspace.GetModules() {
+			if mod != nil {
+				targetFiles, err := mod.TargetFileInfos(ctx)
+				if err != nil {
+					return fmt.Errorf("workspace module target file infos: %w", err)
+				}
+				for _, file := range targetFiles {
+					if file.ModuleIdentity() != nil {
+						directDepsIdentities[file.ModuleIdentity().IdentityString()] = "workspace"
+						break
+					}
+				}
+			}
+		}
+	}
+	b.logger.Debug(
+		"module",
+		zap.Any("module_direct_dependencies", directDepsIdentities),
+	)
+
+	// categorize image files into direct vs transitive dependencies
+	allImgFiles := make(map[string]map[string][]string)    // for logging purposes only, modIdentity:filepath:imports
+	targetFiles := make(map[string]struct{})               // filepath
+	directDepsFilesToModule := make(map[string]string)     // filepath:modIdentity
+	transitiveDepsFilesToModule := make(map[string]string) // filepath:modIdentity
+	for _, file := range builtImage.Files() {
+		{
+			modIdentity := "local"
+			if file.ModuleIdentity() != nil {
+				modIdentity = file.ModuleIdentity().IdentityString()
+			}
+			if _, ok := allImgFiles[modIdentity]; !ok {
+				allImgFiles[modIdentity] = make(map[string][]string)
+			}
+			allImgFiles[modIdentity][file.Path()] = file.FileDescriptor().GetDependency()
+		}
+		if file.ModuleIdentity() != nil && file.IsImport() {
+			if _, ok := directDepsIdentities[file.ModuleIdentity().IdentityString()]; ok {
+				directDepsFilesToModule[file.Path()] = file.ModuleIdentity().IdentityString()
+			} else {
+				transitiveDepsFilesToModule[file.Path()] = file.ModuleIdentity().IdentityString()
+			}
+		} else {
+			// file has no module identity (workspace local), or is no import (is local)
+			targetFiles[file.Path()] = struct{}{}
+		}
+	}
+	b.logger.Debug(
+		"image_files",
+		zap.Any("all_with_imports", allImgFiles),
+		zap.Any("local", targetFiles),
+		zap.Any("from_direct_dependencies", directDepsFilesToModule),
+		zap.Any("from_transitive_dependencies", transitiveDepsFilesToModule),
+	)
+
+	// validate import statements of target files against dependencies categorization above
+	for _, file := range builtImage.Files() {
+		if file.IsImport() {
+			continue // only check import statements in local files
+		}
+		// .GetDependency() returns an array of file path imports in the file descriptor
+		for _, importFilePath := range file.FileDescriptor().GetDependency() {
+			if _, ok := targetFiles[importFilePath]; ok {
+				continue // import comes from local
+			}
+			if _, ok := directDepsFilesToModule[importFilePath]; ok {
+				continue // import comes from direct dep
+			}
+			warnMsg := fmt.Sprintf(
+				"Target proto file %q imports %q, not found in your local target files or direct dependencies",
```

Comment:
```suggestion
				"File %q imports %q, which is not found in your local files or direct dependencies",
```

Reply:
I thought it was important saying that we are looking only target files, not all local files. I'll change.